tool/lib/base-canonical-generator.js

```diff
@@ -0,0 +1,70 @@
+// -*- mode: js; indent-tabs-mode: nil; js-basic-offset: 4 -*-
+//
+// This file is part of Genie
+//
+// Copyright 2020 The Board of Trustees of the Leland Stanford Junior University
+//
+// Author: Silei Xu <silei@cs.stanford.edu>
+//
+// See COPYING for details
+"use strict";
+
+const { clean } = require('../../lib/utils');
+const { pluralize, posTag } = require('../../lib/i18n/american-english');
+
+function genBaseCanonical(canonical, name, ptype) {
```

Comment:
Is this something that we want to run automatically for all Thingpedia devices if one does not specify `#_[canonical]` when uploading?
(Currently, we only run `clean`)

Reply:
I think so - I suppose I should move it to `lib` as you suggested last time?